File: app/frontend/app/models/integration.js
import Ember from 'ember';
import DS from 'ember-data';
import CoughDrop from '../app';
import persistence from '../utils/persistence';

CoughDrop.Integration = DS.Model.extend({
  name: DS.attr('string'),
  user_id: DS.attr('string'),
  custom_integration: DS.attr('boolean'),
  webhook: DS.attr('boolean'),
  render: DS.attr('boolean'),
  render_url: DS.attr('string'),
  user_token: DS.attr('string'),
  button_webhook_url: DS.attr('string'),
  board_render_url: DS.attr('string'),
  insecure_button_webhook_url: function() {
    var url = this.get('button_webhook_url');
    return url && url.match(/^http:/);
  }.property('button_webhook_url'),
  insecure_board_render_url: function() {
    var url = this.get('board_render_url');
    return url && url.match(/^http:/);
  }.property('board_render_url'),
  access_token: DS.attr('string'),
  truncated_access_token: DS.attr('string'),
  displayable_access_token: function() {
    return this.get('access_token') || this.get('truncated_access_token');
  }.property('access_token', 'truncated_access_token'),
  has_multiple_actions: function() {
    return !!(this.get('webhook') && this.get('render'));
  }.property('webhook', 'render'),
  token: DS.attr('string'),
  truncated_token: DS.attr('string'),
  displayable_token: function() {
    return this.get('token') || this.get('truncated_token');
  }.property('token', 'truncated_token'),
});

export default CoughDrop.Integration;

File: lib/json_api/integration.rb
module JsonApi::Integration
  extend JsonApi::Json
  
  TYPE_KEY = 'integration'
  DEFAULT_PAGE = 10
  MAX_PAGE = 25
    
  def self.build_json(obj, args={})
    json = {}
    # TODO: include devices
    
    json['id'] = obj.global_id
    json['name'] = obj.settings['name']
    json['custom_integration'] = !!obj.settings['custom_integration']
    json['webhook'] = !!obj.settings['button_webhook_url']
    json['render'] = !!obj.settings['board_render_url']
    
    if obj.template
      json['integration_key'] = obj.integration_key
      json['template'] = true
    end
    ['icon_url', 'description'].each do |key|
      json[key] = obj.settings[key]
    end
    
    if obj.settings['custom_integration']
      device_token = obj.device.token
      if obj.created_at > 24.hours.ago && obj.device
        json['access_token'] = device_token
        json['token'] = obj.settings['token']
      end
      json['truncated_access_token'] = "...#{device_token[-5, 5]}"
      json['truncated_token'] = "...#{obj.settings['token'][-5, 5]}"
    end

    json
  end
  
  def self.extra_includes(obj, json, args={})
    json['integration']['render_url'] = obj.settings['board_render_url']
    if args[:permissions]
      json['integration']['user_token'] = obj.user_token(args[:permissions])
    end
    json
  end
end
